packages/stream/components/utils.ts

```diff
@@ -0,0 +1,32 @@
+type PendingThenableState = { status: "pending" };
+type FulfilledThenableState<T> = { status: "fulfilled"; value: T };
+type RejectedThenableState = { status: "rejected"; reason: unknown };
+
+type ThenableState<T> =
+  | PendingThenableState
+  | FulfilledThenableState<T>
+  | RejectedThenableState;
+
+export type Thenable<T> = Promise<T> & ThenableState<T>;
+
+export function trackThenableState<T>(promise: Promise<T>): Thenable<T> {
```

Comment:
ref from the other PR: https://github.com/bholmesdev/suspense-from-scratch/pull/1#discussion_r1477345695
i think this works?